rust/src/schema.rs

```diff
@@ -160,3 +160,487 @@ impl Schema {
         &self.fields
     }
 }
+
+/// Error representing a failure while creating the delta log schema.
+#[derive(thiserror::Error, Debug)]
+pub enum DeltaLogSchemaError {
+    /// Error returned when JSON de-serialization of schema components fails.
+    #[error("serde_json::Error: {source}")]
+    JSONSerialization {
+        /// The source serde_json::Error.
+        #[from]
+        source: serde_json::Error,
+    },
+}
+
+/// Factory for creating a Delta log schema for a specific table schema.
+/// REF: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoint-schema
+pub struct DeltaLogSchemaFactory {
+    common_fields: HashMap<String, Vec<SchemaField>>,
+}
+
+impl DeltaLogSchemaFactory {
+    /// Creates a new DeltaLogSchemaFactory which can be used to create Schema's representing the
+    /// Delta log for specific tables.
+    pub fn new() -> Self {
```

Comment:
it looks like this `new` method only creates static values? If so, we could simplify it by declaring these values two lazy statics values:

* a vector of `SchemaField` structs for all fields other than add. 
* fields for add schemafield

Then we can simplify the code by replacing all of this factory struct with a single `delta_log_schema_for_table` function that does the following:

* clone the lazy static vector of non-add `SchemaField`s
* generate add SchemaFIeld based on provided table schema and partition columns
* append generated add SchemaField to cloned SchemaField vector

Reply:
this relates to https://github.com/delta-io/delta-rs/pull/280#discussion_r647880588 as well. i moved the schema creation into `delta_arrow.rs`, skipped both json and delta schema entirely and went straight to arrow and included the lazy statics as recommended.